doc/driver.mld

```diff
@@ -0,0 +1,377 @@
+
+An example of a PPX registered in this phase is
+{{:https://github.com/janestreet/ppx_js_style}ppx_js_style}.
+
+{2 The preprocessing phase}
+
+The preprocessing phase is the first transformation that actually alters the
+AST. In fact, the property of being the "first transformation applied" is what
+defines this phase, and [ppxlib] will thus ensure that only one transformation is
+registered in this phase: it will generate an error otherwise.
+
+An example of a PPX registered in this phase is
+{{:https://github.com/thierry-martinez/metapp}metapp}.
+
+{2 The first instrumentation phase}
+
+This phase is for transformations that {e need} to be run before the
+context-free phase. Historically, it was meant for
+{{:https://en.wikipedia.org/wiki/Instrumentation_(computer_programming)}instrumentation}-related
+PPXs, hence the name. Unlike the {{!"the-preprocessing-phase"}preprocessing
+phase}, registering to this phase provides no guarantee that the transformation
+is run early in the rewriting, as there is no limit in the number of
+transformation registered in this phase, which are then applied in the
+alphabetical order of their name.
+
+If it is not crucial for a transformation to be run before the context-free
+phase, it should be registered to the {{!"the-global-transformation-phase"}global
+transformation phase}.
+
+{2:context-free-phase The context-free phase}
+
+The execution of all registered context-free rules is done in a single top-down
+pass through the AST. Whenever the top-down pass encounter a situation
+triggering rewriting, the corresponding transformation is called. For instance,
+when encountering an extension point corresponding to a rewriting rule, the
+extension point is replaced by the execution of the rule, and the top-down pass
+continues inside the generated code. Similarly, when a deriving attribute is
+found to be attached to a structure or signature item, the result of the
+application of the deriving rule is appended to the AST, and the top-down pass
+continues in the generated code.
+
+Note that the code generation for derivers is applied when "leaving" the AST
+node, that is when all rewriters have been run. Indeed, a deriver like this:
+
+{[
+  type t = [%my_type] [@@deriving deriver_from_type]
+]}
+
+would need the information generated by the [my_type] extender to match on the
+structure of [t].
+
+Also note that in this phase, the execution of the context-free rules are
+intertwined altogether, and it would not make sense to speak about the order of
+application, contrary to the next phase.
+
+{2:global-transfo-phase The global transformation phase}
+
+The global transformation phase is the phase where registered transformations,
+seen as function from and to the {{!Ppxlib.Parsetree}[Parsetree]}, are run. The order on which they are
+applied might matter and change the outcome, but since [ppxlib] knows nothing
+about the transformations, the order applied is alphabetical on the names of the
+transformation.
```

Comment:
typo in "transormation" :)
```suggestion
seen as function from and to the {{!Ppxlib.Parsetree}[Parsetree]}, are run. The applied order might matter and change the outcome, but since [ppxlib] knows nothing
about the transformations, the order applied is alphabetical by the transformation's name.
```